pmd-visualforce/src/test/resources/net/sourceforge/pmd/lang/vf/rule/security/xml/VfUnescapeEl.xml

```diff
@@ -437,21 +437,125 @@
+        console.log({!IF(SomeObj.Field=='1', 'foo', SomeObj.Text__c)});
+    </script>
+</apex:page>
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>CASE() with XSS in conditions is safe</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+<apex:page>
+    <script>
+        console.log({!CASE(SomeObj.Name, SomeObj.text__c, 'res1', SomeObj.Text2__c, 'res2', 'else-res')});
+    </script>
+</apex:page>
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>CASE() with escaped XSS in results is safe</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+<apex:page>
+    <script>
+        console.log({!CASE(SomeObj.Name, 'val1', JSENCODE(SomeObj.Text__c), JSENCODE(SomeObj.Text2__c))});
+    </script>
+</apex:page>
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>CASE() with unescaped XSS in results is unsafe</description>
+        <expected-problems>1</expected-problems>
+        <code><![CDATA[
+<apex:page>
+    <script>
+        console.log({!CASE(SomeObj.Name, 'val1', SomeObj.Text__c, SomeObj.Text2__c)});
+    </script>
+</apex:page>
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>LPAD() is not inherently safe</description>
+        <expected-problems>3</expected-problems>
+        <code><![CDATA[
+<apex:page>
+    <script>
+        console.log({!LPAD(SomeObj.Text__c, 100)});
+        console.log({!LPAD(SomeObj.Text__c, 100, JSENCODE(SomeObj.OtherText__c))});
+        console.log({!LPAD(JSENCODE(SomeObj.Text__c), 100, SomeObj.OtherText__c)});
+    </script>
+</apex:page>
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Escaped LPAD() call is fine</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+<apex:page>
+    <script>
+        console.log({!JSENCODE(LPAD(SomeObj.Text__c, 100))});
+        console.log({!LPAD(JSENCODE(SomeObj.Text__c), 100)});
+        console.log({!LPAD(JSENCODE(SomeObj.Text__c), 100, JSENCODE(SomeObj.OtherText__c))});
+    </script>
+</apex:page>
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>NOT, ISBLANK, BEGINS evaluate to safe booleans</description>
         <expected-problems>0</expected-problems>
         <code><![CDATA[
 <apex:page>
     <script>
         if({!NOT(yes)}) { maskFormEls(); }
         if({!NOT foo(yes)}) { maskFormEls(); }
+        console.log({!ISBLANK(safeXss)});
+        console.log({!BEGINS(safeXss, 'foo')});
     </script>
 </apex:page>
         ]]></code>
     </test-code>
 
     <test-code>
-        <description>JSON.parse method evaluates quoted EL to safe JSON</description>
-        <expected-problems>0</expected-problems>
+        <description>JSON.parse method evaluates quoted EL to unsafe XSS</description>
+        <expected-problems>3</expected-problems>
```

Comment:
This test case now shows, that the rule changed here... I don't know what quoted vs. non-quoted EL means for VF, but before we obviously said, that with quoted EL, no XSS is possible.
@rsoesemann Can you confirm, that the new behavior is actually correct now?

Reply:
I was able to trivially do an XSS injection with this case, which is why I changed it.

Reply:
For reference, this was my script:
```
<apex:page showHeader="true" sidebar="true" standardController="TestObj__c">

	<script>
    var x = JSON.parse('{!TestObj__c.VulnText__c}');
	 </script>
</apex:page>
```
And a record with `{}'); window.location.replace('https://www.google.com');//` in its VulnText__c field was able to do a redirect.

Reply:
Thanks for checking!